src/instructlab/model/train.py

```diff
@@ -893,34 +930,44 @@ def _mtbench(
 
 
 def _evaluate_dir_of_checkpoints(
-    checkpoints_dir: pathlib.Path, eval_func: typing.Callable[..., float]
-) -> tuple[float, pathlib.Path]:
-    """Run eval_func on all model checkpoints in a directory, returning the best performing."""
-
+    eval_func: typing.Callable[..., float],
+    phase_model: EvalPhaseModel,
+    journal: TrainingJournal,
+) -> EvalResult:
+    """Run eval_func on all model checkpoints in a directory."""
     # TODO: parallelize MMLU over available GPUs
 
-    results: list[typing.Tuple[float, pathlib.Path]] = []
-    for ckpt_path in checkpoints_dir.iterdir():
-        if ckpt_path.is_dir():
-            logger.debug(ckpt_path)
-            ckpt_score = eval_func(model=ckpt_path)
-            click.secho(
-                f"CHECKPOINT EVALUATION: {ckpt_path} SCORED {ckpt_score}",
-                fg="red",
-                bg="cyan",
-            )
-            results.append((ckpt_score, ckpt_path))
+    # doing this to avoid removing checkpoints from same list that we're iterating over.
+    checkpoints_todo = list(
+        set(phase_model.checkpoints) - set(phase_model.finished_checkpoints)
+    )
 
-    if not results:
+    if len(checkpoints_todo) == 0:
         raise RuntimeError(
-            "_evaluate_dir_of_checkpoints - No checkpoints were evaluated successfully at %s. No scores were recorded."
-            % checkpoints_dir
+            "_evaluate_dir_of_checkpoints - No checkpoints were evaluated, 'checkpoints_todo' was empty"
+        )
+
+    for checkpoint in checkpoints_todo:
+        logger.debug(str(checkpoint))
+        checkpoint_score = eval_func(model=checkpoint)
```

Comment:
I recommend a try/catch here so we can fail graciously.

Reply:
If there's a failure I'd like that failure to bubble up rather than be caught. I could catch the error, add some context, and then reraise. Would that be preferred?